### DISCUSSION TITLE
Writing a Jest-based integration test against Firestore

### DISCUSSION THREAD

**Discussion by electricmonk** [invertase/react-native-firebase#4719] [Q&A] [answered]
Hi there,

We're writing a React Native app using `react-native-firebase` for a chat component. 
We're attempting to write an integration test of the chat component against Cloud Firestore; when running Jest, we get the `Invariant Violation: Native module cannot be null` issue. I tried adding a `transformIgnorePatterns` and then ran into #4265.

All answers seem to point towards mocking out Firebase - but this is not what we need; our goal is creating a test that runs against a locally-running [Firebase Emulator](https://firebase.google.com/docs/rules/emulator-setup).

As I'm a complete noob in react-native, I'd appreciate any pointers towards how to do this properly.

Thanks!

**Comment by mikehardy** [comment, marked as answer]
Hi there!
Please investigate our own testing infrastructure for demonstrations of either:

1) how to mock it out: https://github.com/invertase/react-native-firebase/blob/master/jest.setup.ts or
2) a clear indication that in order to run it you will have to be running in react-native inside an emulator, and just set your emulator:

https://github.com/invertase/react-native-firebase/blob/36fcfd09df0a8e2660fc55f7cfbf62f1c7aff01a/tests/app.js#L44-L45

**Reply by mikehardy** [reply]
Also maybe of interest:

1) CI emulator setup https://github.com/invertase/react-native-firebase/blob/36fcfd09df0a8e2660fc55f7cfbf62f1c7aff01a/.github/workflows/tests_e2e.yml#L43-L54

2) CI emulator configuraiton https://github.com/invertase/react-native-firebase/tree/master/.github/workflows/scripts

3) emulator run scripts https://github.com/invertase/react-native-firebase/blob/36fcfd09df0a8e2660fc55f7cfbf62f1c7aff01a/package.json#L24-L25

**Comment by electricmonk** [comment]
Thanks @mikehardy 
My intent is to TDD the chat system, which would benefit from short feedback circles; an emulator obviously makes this more complex.
Is it indeed impossible to run the code from Node.js? if so, I'd appreciate a pointer to why this is so...

I'll look at your tests anyway - thanks!

**Reply by mikehardy** [reply]
So, this library wraps firebase-android-sdk which is all native java code that runs on an android device. The emulator is used through native code. How would you expect to do that without it running on an emulator 🤔 

Note that react-native has something called "fast refresh" where the built-in metro bundler will detect any changes to a component  source file and if it sees them it will reload the component.

So if you were to, for instance, have an "E2E" test component, and you loaded it on for instance e2e.jsx, and in that component you configured the use of the emulator and then had a series of tests, you would find that you could make a small change while the app was running in the emulator, and approximately 2 seconds later more or less it would refresh the component, and re-run the tests.

You should try it

**Reply by electricmonk** [reply]
Thanks!

Just out of curiosity (and possibly ignorance), is there any reason not to use plain Javascript to interact with Firestore? like you would on a browser-side application?

**Reply by mikehardy** [reply]
native networking power (think, flaky networks, application lifecycle listeners, auto reconnects etc), native filesystem caching power. But sure, do as you wish, firebase-js-sdk works for most modules. I won't be available to answer more questions sorry - play with it and learn, good luck with your project

**Reply by electricmonk** [reply]
Thanks Mike!
